File: apps/api/app/models/ranked.py
"""Pydantic request/response models for Phase 4.0 ranked duels.

No internal database IDs are exposed beyond opaque UUIDs already used as
primary keys elsewhere in the API (matches the existing convention — game_id,
challenge tokens — of treating server-generated UUIDs as the public identity
rather than inventing a second opaque-ID layer).
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class RankedQueueInfo(BaseModel):
    mode: str
    label: str
    queue_version: str
    rating_algorithm_version: str
    placement_count: int


class RankedQueuesResponse(BaseModel):
    queues: list[RankedQueueInfo]
    ranked_enabled: bool
    matchmaking_enabled: bool


class JoinQueueResponse(BaseModel):
    status: str            # "waiting" | "matched"
    mode: str
    queue_entry_id: Optional[str] = None
    match_id: Optional[str] = None


class MatchmakingStatusResponse(BaseModel):
    status: str             # "not_in_queue" | "waiting" | "matched" | "cancelled"
    mode: str
    waited_seconds: Optional[float] = None
    match_id: Optional[str] = None


class RankedParticipantPublic(BaseModel):
    status: str              # board_ready | in_progress | complete | awaiting_opponent | abandoned | protected_abort
    game_id: Optional[str] = None


class RankedMatchPublic(BaseModel):
    match_id: str
    mode: str
    status: str
    settlement_status: str
    deadline: str
    you: RankedParticipantPublic
    opponent_status: str    # "hidden" (pre-settlement) | the opponent's status (post-settlement)


class RatingChange(BaseModel):
    prior_rating: float
    new_rating: float
    delta: float
    prior_rd: float
    new_rd: float


class RankedSettlementView(BaseModel):
    match_id: str
    outcome: str                       # "win" | "loss" | "draw" (from the viewer's perspective)
    your_score: float
    opponent_score: float
    tie_break_used: Optional[str] = None
    rating_change: RatingChange
    placement_progress: Optional[str] = None    # e.g. "Placement 3 of 7" or None once established
    division_change: Optional[str] = None       # e.g. "Rotation -> Starter" only if applicable
    settled_at: str


class PendingSettlementResponse(BaseModel):
    status: str   # "awaiting_opponent" | "settled"
    match_id: str


class QueueRatingResponse(BaseModel):
    mode: str
    established: bool
    rating: Optional[float] = None       # hidden (None) while in placement, per spec K
    rd: Optional[float] = None
    uncertainty_label: str               # human-readable: "still calibrating" | "provisional" | "established"
    valid_rated_matches: int
    division: Optional[str] = None       # only shown once established


class PlacementStateResponse(BaseModel):
    mode: str
    valid_matches_completed: int
    required_matches: int
    established: bool


class RatingHistoryEntry(BaseModel):
    match_id: str
    outcome: str
    pre_rating: float
    post_rating: float
    delta: float
    created_at: str


class RatingHistoryResponse(BaseModel):
    mode: str
    entries: list[RatingHistoryEntry]


class LeaderboardEntry(BaseModel):
    rank: int
    owner_sub: str          # stable pagination key only — never a competitive factor
    rating: float
    rd: float
    division: Optional[str] = None


class LeaderboardResponse(BaseModel):
    mode: str
    enabled: bool
    entries: list[LeaderboardEntry]
    next_cursor: Optional[str] = None
    updated_at: str
    queue_version: str
    rating_algorithm_version: str


class SurroundingRankResponse(BaseModel):
    mode: str
    your_rank: Optional[int] = None
    entries: list[LeaderboardEntry]


class RankedReadinessResponse(BaseModel):
    readiness_level: str
    ranked_enabled: bool
    matchmaking_enabled: bool
    rating_writes_enabled: bool
    public_leaderboard_enabled: bool
    rating_algorithm_version: str
    queue_versions: dict[str, str]
    pending_match_count: int
    pending_rating_count: int
    last_successful_settlement_at: Optional[str] = None

File: apps/web/src/types/ranked.ts
/**
 * Types for Phase 4.0 ranked duels. Server-shape interfaces mirror
 * apps/api/app/models/ranked.py field-for-field (snake_case, matching the
 * FastAPI/Pydantic response — same convention as types/draft.ts).
 */
import type { DraftGameState } from "./draft";

export type RankedMode = "apex_1y" | "prime_3y" | "foundation_5y";

export const RANKED_MODES: RankedMode[] = ["apex_1y", "prime_3y", "foundation_5y"];

export const RANKED_MODE_LABELS: Record<RankedMode, string> = {
  apex_1y: "1Y Apex",
  prime_3y: "3Y Prime",
  foundation_5y: "5Y Foundation",
};

// ── Queues ──────────────────────────────────────────────────────────────

export interface RankedQueueInfo {
  mode: RankedMode;
  label: string;
  queue_version: string;
  rating_algorithm_version: string;
  placement_count: number;
}

export interface RankedQueuesResponse {
  queues: RankedQueueInfo[];
  ranked_enabled: boolean;
  matchmaking_enabled: boolean;
}

export interface JoinQueueResponse {
  status: "waiting" | "matched";
  mode: RankedMode;
  queue_entry_id: string | null;
  match_id: string | null;
}

export interface MatchmakingStatusResponse {
  status: "not_in_queue" | "waiting" | "matched" | "cancelled";
  mode: RankedMode;
  waited_seconds: number | null;
  match_id: string | null;
}

// ── Match + game ────────────────────────────────────────────────────────

export type RankedParticipantStatus =
  | "board_ready"
  | "in_progress"
  | "complete"
  | "awaiting_opponent"
  | "abandoned"
  | "protected_abort";

export interface RankedMatchPublic {
  match_id: string;
  mode: RankedMode;
  status: string;
  settlement_status: string;
  deadline: string;
  you: { status: RankedParticipantStatus; game_id: string | null };
  opponent_status: string; // "hidden" pre-settlement
}

// The ranked game itself reuses the exact Draft game-state shape — same
// board/round/offer contract, same hidden-information guarantees.
export type RankedGameState = DraftGameState;

// ── Settlement ──────────────────────────────────────────────────────────

export interface RatingChange {
  prior_rating: number;
  new_rating: number;
  delta: number;
  prior_rd: number;
  new_rd: number;
}

export interface RankedSettlementView {
  match_id: string;
  outcome: "win" | "loss" | "draw";
  your_score: number;
  opponent_score: number;
  tie_break_used: string | null;
  rating_change: RatingChange;
  placement_progress: string | null;
  division_change: string | null;
  settled_at: string;
}

export interface PendingSettlementResponse {
  status: "awaiting_opponent" | "settled";
  match_id: string;
}

export type RankedSettlementOrPending = RankedSettlementView | PendingSettlementResponse;

export function isSettled(r: RankedSettlementOrPending): r is RankedSettlementView {
  return "outcome" in r;
}

// ── Rating / placement / history ───────────────────────────────────────

export interface QueueRatingResponse {
  mode: RankedMode;
  established: boolean;
  rating: number | null; // hidden during placements
  rd: number | null;
  uncertainty_label: string;
  valid_rated_matches: number;
  division: string | null;
}

export interface PlacementStateResponse {
  mode: RankedMode;
  valid_matches_completed: number;
  required_matches: number;
  established: boolean;
}

export interface RatingHistoryEntry {
  match_id: string;
  outcome: "win" | "loss" | "draw";
  pre_rating: number;
  post_rating: number;
  delta: number;
  created_at: string;
}

export interface RatingHistoryResponse {
  mode: RankedMode;
  entries: RatingHistoryEntry[];
}

// ── Leaderboard ─────────────────────────────────────────────────────────

export interface LeaderboardEntry {
  rank: number;
  owner_sub: string;
  rating: number;
  rd: number;
  division: string | null;
}

export interface LeaderboardResponse {
  mode: RankedMode;
  enabled: boolean;
  entries: LeaderboardEntry[];
  next_cursor: string | null;
  updated_at: string;
  queue_version: string;
  rating_algorithm_version: string;
}

export interface SurroundingRankResponse {
  mode: RankedMode;
  your_rank: number | null;
  entries: LeaderboardEntry[];
}

// ── Readiness ───────────────────────────────────────────────────────────

export interface RankedReadinessResponse {
  readiness_level: "disabled" | "simulation_only" | "internal_alpha" | "closed_alpha" | "public_beta";
  ranked_enabled: boolean;
  matchmaking_enabled: boolean;
  rating_writes_enabled: boolean;
  public_leaderboard_enabled: boolean;
  rating_algorithm_version: string;
  queue_versions: Record<string, string>;
  pending_match_count: number;
  pending_rating_count: number;
  last_successful_settlement_at: string | null;
}

// ── Local UI state ──────────────────────────────────────────────────────

export type RankedUIPhase =
  | "loading"
  | "queue_idle"
  | "queue_waiting"
  | "matched"
  | "playing"
  | "awaiting_opponent"
  | "settled"
  | "error";

export interface RankedUIState {
  phase: RankedUIPhase;
  mode: RankedMode | null;
  matchId: string | null;
  gameState: RankedGameState | null;
  settlement: RankedSettlementView | null;
  errorMessage: string | null;
  waitedSeconds: number;
}

export type RankedUIAction =
  | { type: "SET_MODE"; mode: RankedMode }
  | { type: "QUEUE_JOINED"; result: JoinQueueResponse }
  | { type: "QUEUE_WAIT_TICK"; waitedSeconds: number }
  | { type: "QUEUE_CANCELLED" }
  | { type: "MATCHED"; matchId: string }
  | { type: "GAME_LOADED"; gameState: RankedGameState }
  | { type: "AWAITING_OPPONENT" }
  | { type: "SETTLED"; settlement: RankedSettlementView }
  | { type: "SET_ERROR"; message: string }
  | { type: "RESET" };
